draft-ietf-quic-transport.md

```diff
@@ -2697,6 +2697,43 @@ also be forward-secure encrypted.  Since the attacker will not have the forward
 secure key, the attacker will not be able to generate forward-secure encrypted
 packets with ACK frames.
 
+## Stream fragmentation and reassembly attacks
+
+An adversarial client may attempt to
+exhaust server memory resource by performing
+a stream fragmentation and reassembly attack, similar to the UDP/ICMP
+"Teardrop" fragmentation attacks. The adversarial client would open a stream,
+and send some STREAM DATA packets containing fragments of the stream content.
+The goal of the attack is to induce the receiving implementation to commit
+memory buffers while waiting that the stream data can be reassembled.
+
+For example, a client
+might send some octets at the beginning of a
+stream and some octets much further
+away in the stream. A vulnerable server, on receiving the stream data fragments,
+might allocate memory covering the whole space from beginning to end. The client
+would then repeat the process on a large number of streams and a large number
+of connections, inducing the server to commit a large amout of memory.
+
+This attack can be mitigated by not
+committing memory for stream data reassembly,
+and simply keeping the STREAM DATA frames until enough fragments have been
+received and the data can be delivered to the application in proper sequence.
```

Comment:
As you say, Mikkel, "it is hard to provide general advice". So maybe that's own we should rewrite the "advice" part of this PR. Something like "It is hard to provide general advice. QUIC deployments SHOULD provide mitigations against the stream fragmentation attack, which MAY be avoiding over-committing memory, delaying reassembly of STREAM DATA frames, or implementing heuristics based the age and duration of reassembly holes."